File: projects/src/main/java/io/nirahtech/erp/projects/WorkPackage.java
package io.nirahtech.erp.projects;

import java.io.File;
import java.io.Serializable;
import java.time.Duration;
import java.time.LocalDate;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import io.nirahtech.erp.projects.utils.CriticalPathCalculator;

public final class WorkPackage implements Serializable {
	private final WorkPackageIdentifier id;
	private final String name;
	private Duration duration;
	private String description = null;
	private ProjectMember referent = null;
	private LocalDate startDate = LocalDate.now();
	private Status status = Status.IDLE;
	private final Set<File> documentations = new HashSet<>();
	private final Set<String> labels = new HashSet<>();
	private final Set<Task> tasks = new HashSet<>();
	private final Set<Milestone> milestones = new HashSet<>();
	private final Map<Status, Runnable> eventListeners = new HashMap<>();

	public WorkPackage(
			final WorkPackageIdentifier id,
			final String name,
			final String description,
			final ProjectMember referent,
			final LocalDate startDate,
			final Duration duration,
			final Status status,
			final Set<File> documentations,
			final Set<String> labels,
			final Set<Task> tasks,
			final Set<Milestone> milestones) {
		this.id = id;
		this.name = name;
		this.description = description;
		this.referent = referent;
		this.startDate = startDate;
		this.duration = duration;
		this.status = status;
		this.documentations.addAll(documentations);
		this.labels.addAll(labels);
		this.tasks.addAll(tasks);
		this.milestones.addAll(milestones);
	}

	public String getDescription() {
		return description;
	}
	public void setDescription(String description) {
		this.description = description;
	}
	public void setReferent(ProjectMember referent) {
		this.referent = referent;
	}
	public void setStartDate(LocalDate startDate) {
		this.startDate = startDate;
	}
	public void setStatus(Status status) {
		this.status = status;
		if (this.eventListeners.containsKey(status)) {
			Runnable callback = this.eventListeners.get(status);
			callback.run();
		}
	}
	public void addTask(final Task task) {
		this.tasks.add(task);
	}
	public void addTasks(final Task... tasks) {
		this.tasks.addAll(Set.of(tasks));
	}
	public void addLabel(final String label) {
		this.labels.add(label);
	}
	public void addLabels(final String... labels) {
		this.labels.addAll(Set.of(labels));
	}
	public void addDocumentation(final File documentation) {
		this.documentations.add(documentation);
	}
	public void addDocumentations(final File... documentations) {
		this.documentations.addAll(Set.of(documentations));
	}
	
	public void addMilestone(final Milestone milestone) {
		this.milestones.add(milestone);
	}
	public void addMilestones(final Milestone... milestones) {
		this.milestones.addAll(Set.of(milestones));
	}
	public Collection<File> getDocumentations() {
		return documentations;
	}

	public Duration getDuration() {
		return duration;
	}

	public WorkPackageIdentifier getId() {
		return id;
	}

	public Collection<String> getLabels() {
		return labels;
	}

	public Collection<Milestone> getMilestones() {
		return milestones;
	}

	public String getName() {
		return name;
	}

	public ProjectMember getReferent() {
		return referent;
	}

	public LocalDate getStartDate() {
		return startDate;
	}

	public Status getStatus() {
		return status;
	}

	public Collection<Task> getTasks() {
		return tasks;
	}

	public LocalDate getDeadLine() {
		return this.startDate.plus(this.duration);
	}
	public void setDuration(Duration duration) {
		this.duration = duration;
	}

	public Optional<LocalDate> getEstimatedEndDate() {
		Optional<LocalDate> estimatedEndDate = Optional.empty();
		List<Task> list = this.tasks.stream().toList();
		if (!list.isEmpty()) {
			Task mostFutureTask = list.get(0);

			for (Task task : list) {
				if (task.getDeadLine().isAfter(mostFutureTask.getDeadLine())) {
					mostFutureTask = task;
				}
			}

			estimatedEndDate = Optional.ofNullable(mostFutureTask.getDeadLine());
		}
		return estimatedEndDate;
	}

	public final long computeCumulativeWorkingDays() {
		return tasks.stream()
				.mapToLong(task -> task.getDuration().toDays())
				.sum();
	}

	public final long computeMaxCriticalPathValue() {
		return CriticalPathCalculator.compute(this);
	}

	public void addEventListenerOnStatusChanged(final Status status, final Runnable callback) {
		this.eventListeners.put(status, callback);
	}
}

File: webapp/src/main/react/src/data/WorkPackage.ts
import Duration from "./Duration";
import LocalDate from "./LocalDate";
import { Milestone } from "./Milestone";
import ProjectMember from "./ProjectMember";
import Status from "./Status";
import { Task } from "./Task";
import WorkPackageIdentifier from "./WorkPackageIdentifier";

class WorkPackage {
  private readonly id: WorkPackageIdentifier;
  private readonly name: string;
  private description: string | null = null;
  private referent: ProjectMember | null = null;
  private startDate: LocalDate = LocalDate.now();
  private status: Status = Status.IDLE;
  private duration: Duration;
  private readonly documentations: Set<File> = new Set<File>();
  private readonly labels: Set<string> = new Set<string>();
  private readonly tasks: Set<Task> = new Set<Task>();
  private readonly milestones: Set<Milestone> = new Set<Milestone>();
  private readonly eventListeners: Map<Status, Function> = new Map<Status, Function>();

  constructor(
    id: WorkPackageIdentifier,
    name: string,
    description: string | null,
    referent: ProjectMember | null,
    startDate: LocalDate,
    duration: Duration,
    status: Status,
    documentations: Set<File>,
    labels: Set<string>,
    tasks: Set<Task>,
    milestones: Set<Milestone>
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.referent = referent;
    this.startDate = startDate;
    this.duration = duration;
    this.status = status;
    this.documentations = new Set<File>(documentations);
    this.labels = new Set<string>(labels);
    this.tasks = new Set<Task>(tasks);
    this.milestones = new Set<Milestone>(milestones);
    this.eventListeners = new Map<Status, Function>();
  }

  getDescription(): string | null {
    return this.description;
  }

  setDescription(description: string | null): void {
    this.description = description;
  }

  setReferent(referent: ProjectMember | null): void {
    this.referent = referent;
  }

  setStartDate(startDate: LocalDate): void {
    this.startDate = startDate;
  }

  setStatus(status: Status): void {
    this.status = status;
    if (this.eventListeners.has(status)) {
      const callback = this.eventListeners.get(status);
      if (callback) {
        callback();
      }
    }
  }

  addTask(task: Task): void {
    this.tasks.add(task);
  }

  addTasks(tasks: Task[]): void {
    tasks.forEach((task) => {
      this.tasks.add(task);
    });
  }

  addLabel(label: string): void {
    this.labels.add(label);
  }

  addLabels(labels: string[]): void {
    labels.forEach((label) => {
      this.labels.add(label);
    });
  }

  addDocumentation(documentation: File): void {
    this.documentations.add(documentation);
  }

  addDocumentations(documentations: File[]): void {
    documentations.forEach((documentation) => {
      this.documentations.add(documentation);
    });
  }

  addMilestone(milestone: Milestone): void {
    this.milestones.add(milestone);
  }

  addMilestones(milestones: Milestone[]): void {
    milestones.forEach((milestone) => {
      this.milestones.add(milestone);
    });
  }

  getDocumentations(): Set<File> {
    return this.documentations;
  }

  getDuration(): Duration {
    return this.duration;
  }

  getId(): WorkPackageIdentifier {
    return this.id;
  }

  getLabels(): Set<string> {
    return this.labels;
  }

  getMilestones(): Set<Milestone> {
    return this.milestones;
  }

  getName(): string {
    return this.name;
  }

  getReferent(): ProjectMember | null {
    return this.referent;
  }

  getStartDate(): LocalDate {
    return this.startDate;
  }

  getStatus(): Status {
    return this.status;
  }

  getTasks(): Set<Task> {
    return this.tasks;
  }

  getDeadLine(): LocalDate {
    return this.startDate.plus(this.duration);
  }

  setDuration(duration: Duration): void {
    this.duration = duration;
  }

  getEstimatedEndDate(): LocalDate | null {
    let mostFutureTask: Task | undefined = undefined;

    for (const task of Array.from(this.tasks)) {
      if (!mostFutureTask || task.getDeadLine().isAfter(mostFutureTask.getDeadLine())) {
        mostFutureTask = task;
      }
    }

    return mostFutureTask ? mostFutureTask.getDeadLine() : null;
  }

  computeCumulativeWorkingDays(): number {
    return Array.from(this.tasks).reduce((totalDays, task) => totalDays + task.getDuration().toDays(), 0);
  }

  computeMaxCriticalPathValue(): number {
    // Calculate critical path value for this work package
    return 0; // Placeholder, replace with actual calculation
  }

  addEventListenerOnStatusChanged(status: Status, callback: Function): void {
    this.eventListeners.set(status, callback);
  }
}

export { WorkPackage };
